### PR TITLE
Fixes #22374: fix route table update in ec2_vpc_route_table module

### LINKED ISSUE #22374: ec2_vpc_route_table can't update routes
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME

ec2_vpc_route_table

##### ANSIBLE VERSION
```
ansible 2.1.3.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### OS / ENVIRONMENT

N/A

##### SUMMARY

Ran a script to create a NAT instance, created the routes going through the NAT using ec2_vpc_route_table 
Deleted the NAT and ran the same script again.
ec2_vpc_route_table was not able to update the route with the new instance id, but left the old network interface (which no longer existed) in place, thereby resulting in a black hole.

##### STEPS TO REPRODUCE

```yaml
- name: Create Backend route 1 and route it through NAT 1
  ec2_vpc_route_table:
    vpc_id: '{{ vpc_id }}'
    region: '{{ vpc_region }}'
    tags:
      Name: "{{ vpc_name }} Backend network 1"
    routes:
      - dest: 0.0.0.0/0
        instance_id: '{{ instance_id }}'
    subnets:
      - "{{ vpc_subnet['web_subnet']['subnet_one'].resource_tags.Name }}"
      - "{{ vpc_subnet['db_subnet']['subnet_one'].resource_tags.Name }}"

```

##### EXPECTED RESULTS

I naturally expected that the route table would be updated

##### ACTUAL RESULTS

```
changed: [10.77.200.10] => {"changed": true, "invocation": {"module_args": {"aws_access_key": null, "aws_secret_key": null, "ec2_url": null, "lookup": "tag", "profile": null, "propagating_vgw_ids": null, "region": "us-west-2", "route_table_id": null, "routes": [{"destination_cidr_block": "0.0.0.0/0", "instance_id": "i-1234567890123456"}], "security_token": null, "state": "present", "subnets": ["test - web - us-west-2c",  "test - database - us-west-2c"], "tags": {"Name": "test Backend network 1"}, "validate_certs": true, "vpc_id": "vpc-12345678"}, "module_name": "ec2_vpc_route_table"}, "route_table": {"id": "rtb-23456789", "routes": [{"destination_cidr_block": "10.99.0.0/16", "gateway_id": null, "instance_id": "i-0987654321098765", "interface_id": "eni-12345678", "origin": "CreateRoute", "state": "active", "vpc_peering_connection_id": null}, {"destination_cidr_block": "10.77.0.0/16", "gateway_id": "local", "instance_id": null, "interface_id": null, "origin": "CreateRouteTable", "state": "active", "vpc_peering_connection_id": null}, {"destination_cidr_block": "0.0.0.0/0", "gateway_id": null, "instance_id": null, "interface_id": "eni-87654321", "origin": "CreateRoute", "state": "blackhole", "vpc_peering_connection_id": null}], "tags": {"Name": "test Backend network 1"}, "vpc_id": "vpc-12345678"}}

```

### PR DESCRIPTION
**Pull request by Erouan50**
Fixes #22374

##### SUMMARY
The ec2_vpc_route_table module notifies about a change on the route table when the instance Id of the NAT instance has changed, but in fact, nothing changes. The module call the create_route function the AWS SDK to add a new route with the same cidr. The AWS SDK should return an error instead of nothing.

Call replace_route function instead of create_route when a route table with the same cidr but with different target destination is present.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
modues/cloud/amazon/ec2_vpc_route_table.py

##### ANSIBLE VERSION
```
ansible 2.4.0 (fix-ec2_vpc_route_table 57bf98aeae) last updated 2017/07/24 08:49:28 (GMT -400)
python version = 2.7.13 (default, Jun 26 2017, 10:20:05) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]

```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/ec2_vpc_route_table.py [hunk 1]
```diff
@@ -392,6 +402,15 @@ def ensure_routes(vpc_conn, route_table, route_specs, propagating_vgw_ids,
                 if e.error_code == 'DryRunOperation':
                     pass
 
+        for route_spec in route_specs_to_recreate:
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_route_table.py:405]: Above this line can you please do:
`if not check_mode:`

File: lib/ansible/modules/cloud/amazon/ec2_vpc_route_table.py [hunk 2]
```diff
@@ -392,6 +402,15 @@ def ensure_routes(vpc_conn, route_table, route_specs, propagating_vgw_ids,
...
+            try:
+                vpc_conn.replace_route(route_table.id,
+                                       dry_run=check_mode,
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_route_table.py]: Please delete this line.

File: lib/ansible/modules/cloud/amazon/ec2_vpc_route_table.py [hunk 3]
```diff
@@ -392,6 +402,15 @@ def ensure_routes(vpc_conn, route_table, route_specs, propagating_vgw_ids,
                 if e.error_code == 'DryRunOperation':
                     pass
 
+        for route_spec in route_specs_to_recreate:
+            try:
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_route_table.py]: You can delete this line (and the `except....`) as EC2ResponseError triggered by this function will be caught in main()

**Review by s-hertel — CHANGES_REQUESTED** [review]
This looks good to me. However, the fact that DryRun is being utilized in the module is a mistake. I have a pull request https://github.com/ansible/ansible/pull/26591 to remove all traces. This fix is smaller and probably easier to do a quick merge before mine (and I don't mind rebasing mine), but I'd prefer not to add more of the issue into this module.

**Comment by Erouan50** [comment]
I just did the changes. Is it ok?

**Review by s-hertel — APPROVED** [review]
LGTM!

**Comment by s-hertel** [comment]
@ryansb @willthames can I get another review on this?

File: lib/ansible/modules/cloud/amazon/ec2_vpc_route_table.py (hunk 1 above)
Comment (ryansb) [reply lib/ansible/modules/cloud/amazon/ec2_vpc_route_table.py:405]: Addressed.

**Review by ryansb — APPROVED** [review]

**Comment by ryansb** [comment]
rebuild_merge

**Comment by ryansb** [comment]
This would be good to backport to 2.4 as well, since it's a fix to how changes are calculated.

**Comment by willthames** [comment]
This change doesn't appear to fix #22734.

I run my playbook, delete the NAT instance, rerun the playbook with this change and still see

```
TASK [ec2_vpc : create outbound route table] ****************************************************************************************************************************************************
task path: /Users/will/src/ansible-cf-tf-playoff/ansible/roles/ec2_vpc/tasks/main.yml:59
Friday 08 September 2017  15:16:03 -0700 (0:00:00.037)       0:00:21.219 ****** 
ok: [vpc-runner] => {"changed": false, "failed": false, "route_table": {"id": "rtb-06153b62", "routes": [{"destination_cidr_block": "10.3.0.0/16", "gateway_id": "local", "instance_id": null, "interface_id": null, "origin": "CreateRouteTable", "state": "active", "vpc_peering_connection_id": null}, {"destination_cidr_block": "0.0.0.0/0", "gateway_id": null, "instance_id": null, "interface_id": null, "origin": "CreateRoute", "state": "blackhole", "vpc_peering_connection_id": null}], "tags": {"Name": "private"}, "vpc_id": "vpc-fe871c9a"}}
```

The reason for this is [actually documented in the relevant code](https://github.com/Erouan50/ansible/blob/5fd23529ed1cf47d1e1f9902ad7fe6b43c171f1a/lib/ansible/modules/cloud/amazon/ec2_vpc_route_table.py#L319-L324) around this

```
    # This is a workaround to catch managed NAT gateways as they do not show
    # up in any of the returned values when describing route tables.
    # The caveat of doing it this way is that if there was an existing
    # route for another nat gateway in this route table there is not a way to
    # change to another nat gateway id. Long term solution would be to utilise
    # boto3 which is a very big task for this module or to update boto.
```

**Comment by willthames** [comment]
I have no problem with the code in this PR, I'm just not sure what it achieves differently to the current status.

**Comment by s-hertel** [comment]
@willthames In his [commit](https://github.com/Erouan50/ansible/commit/75479f5a0d1fc5fe410fc30f13c9551c645ff40a) he referenced https://github.com/ansible/ansible/issues/26664 but used https://github.com/ansible/ansible/issues/22374 for the title.

**Pull request merged**